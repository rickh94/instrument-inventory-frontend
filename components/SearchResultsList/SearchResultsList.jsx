import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import { Table, TableRow, TableBody, TableCell, TableHead } from '@material-ui/core'
import { titleCase } from '../../libs/titleCase'

export const ResultRowInternal = ({ item, history }) => (
  <TableRow
    hover
    onClick={() => history.push(`/instrument/${item.id}`)}
    component="tr"
    data-testid="result-row"
  >
    <TableCell component="td" data-testid="type-and-number">
      {titleCase(item.fields['Instrument Type'])} {item.fields.Number}
    </TableCell>
    <TableCell component="td" data-testid="size">
      {item.fields.Size}
    </TableCell>
    <TableCell component="td" data-testid="assignedTo">
      {item.fields['Assigned To'] || 'Not Assigned'}
    </TableCell>
    <TableCell component="td" data-testid="location">
      {item.fields['Location']}
    </TableCell>
  </TableRow>
)

ResultRowInternal.propTypes = {
  item: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
}

const ResultRow = withRouter(ResultRowInternal)

const SearchResultsList = ({ results }) => (
  <Table component="table">
    <TableHead component="thead">
      <TableRow component="tr">
        <TableCell component="th">Instrument</TableCell>
        <TableCell component="th">Size</TableCell>
        <TableCell component="th">Assigned To</TableCell>
        <TableCell component="th">Location</TableCell>
      </TableRow>
    </TableHead>
    <TableBody component="tbody">
      {results.map(item => (
        <ResultRow item={item} key={item.id} />
      ))}
    </TableBody>
  </Table>
)

SearchResultsList.propTypes = {
  results: PropTypes.array.isRequired,
}

export default SearchResultsList
