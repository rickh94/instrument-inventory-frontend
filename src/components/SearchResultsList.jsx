import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import { Table, TableRow, TableBody, TableCell, TableHead } from '@material-ui/core'
import { titleCase } from '../libs/titleCase'

export const ResultRowInternal = ({ item, history }) => (
  <TableRow
    hover
    onClick={() => history.push(`/instrument/${item.id}`)}
    component="tr"
    data-testid="result-row"
  >
    <TableCell component="td" data-testid="type-and-number">
      {titleCase(item.type)} {item.number}
    </TableCell>
    <TableCell component="td" data-testid="size">
      {item.size}
    </TableCell>
    <TableCell component="td" data-testid="assignedTo">
      {item.assignedTo || 'Not Assigned'}
    </TableCell>
    <TableCell component="td" data-testid="location">
      {item.location}
    </TableCell>
  </TableRow>
)

ResultRowInternal.propTypes = {
  item: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
}

const ResultRow = withRouter(ResultRowInternal)

const SearchResultsList = ({ results }) => (
  <Table component="table">
    <TableHead component="thead">
      <TableRow component="tr">
        <TableCell component="th">Instrument</TableCell>
        <TableCell component="th">Size</TableCell>
        <TableCell component="th">Assigned To</TableCell>
        <TableCell component="th">Location</TableCell>
      </TableRow>
    </TableHead>
    <TableBody component="tbody">
      {results.map(item => (
        <ResultRow item={item} key={item.id} />
      ))}
    </TableBody>
  </Table>
)

SearchResultsList.propTypes = {
  results: PropTypes.array.isRequired,
}

export default SearchResultsList
