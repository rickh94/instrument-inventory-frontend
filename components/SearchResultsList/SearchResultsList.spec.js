import React from 'react'
import SearchResultsList, { ResultRowInternal } from './SearchResultsList'
import { render, fireEvent } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import 'jest-dom/extend-expect'

const testItem = {
  id: 'testid',
  fields: {
    'Instrument Type': 'test',
    Size: '4/4',
    'Assigned To': 'test person',
    Location: 'test location',
    Number: '1-001',
  },
}

const makeTestItem = id => ({
  id: id,
  fields: {
    'Instrument Type': 'test',
    Size: '4/4',
    'Assigned To': `test${id}`,
    Location: 'test',
    Number: '1-001',
  },
})

describe('ResultRowInternal', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <ResultRowInternal item={testItem} history={{ push: jest.fn() }} />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders data from props', () => {
    const { container } = render(
      <ResultRowInternal item={testItem} history={{ push: jest.fn() }} />
    )
    expect(
      container.querySelector('td[data-testid="type-and-number"]')
    ).toHaveTextContent('Test 1-001')
    expect(container.querySelector('td[data-testid="size"]')).toHaveTextContent('4/4')
    expect(container.querySelector('td[data-testid="assignedTo"]')).toHaveTextContent(
      'test person'
    )
    expect(container.querySelector('td[data-testid="location"]')).toHaveTextContent(
      'test location'
    )
  })

  it('goes to instrument page on click', () => {
    const push = jest.fn()
    const { container } = render(
      <ResultRowInternal item={testItem} history={{ push }} />
    )

    fireEvent.click(container.querySelector('tr'))
    expect(push).toHaveBeenCalledWith('/instrument/testid')
  })
})

describe('<SearchResultsList />', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchResultsList
          results={[makeTestItem(0), makeTestItem(1), makeTestItem(2)]}
        />
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('renders a result row for each item', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchResultsList
          results={[makeTestItem(0), makeTestItem(1), makeTestItem(2)]}
        />
      </MemoryRouter>
    )
    expect(container.querySelectorAll('tr[data-testid="result-row"]').length).toBe(3)
  })
})
