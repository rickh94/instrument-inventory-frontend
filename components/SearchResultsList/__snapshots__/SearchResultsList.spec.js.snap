// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultsList /> matches snapshot 1`] = `
<div>
  <table
    class="MuiTable-root"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Instrument
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Size
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Assigned To
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Location
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
        data-testid="result-row"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="type-and-number"
        >
          Test
           
          1-001
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="size"
        >
          4/4
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="assignedTo"
        >
          test0
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="location"
        >
          test
        </td>
      </tr>
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
        data-testid="result-row"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="type-and-number"
        >
          Test
           
          1-001
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="size"
        >
          4/4
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="assignedTo"
        >
          test1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="location"
        >
          test
        </td>
      </tr>
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
        data-testid="result-row"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="type-and-number"
        >
          Test
           
          1-001
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="size"
        >
          4/4
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="assignedTo"
        >
          test2
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
          data-testid="location"
        >
          test
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ResultRowInternal matches snapshot 1`] = `
<div>
  <tr
    class="MuiTableRow-root MuiTableRow-hover"
    data-testid="result-row"
  >
    <td
      class="MuiTableCell-root"
      data-testid="type-and-number"
    >
      Test
       
      1-001
    </td>
    <td
      class="MuiTableCell-root"
      data-testid="size"
    >
      4/4
    </td>
    <td
      class="MuiTableCell-root"
      data-testid="assignedTo"
    >
      test person
    </td>
    <td
      class="MuiTableCell-root"
      data-testid="location"
    >
      test location
    </td>
  </tr>
</div>
`;
